### DISCUSSION TITLE
Deprecating TLS 1.0/1.1 protocols

### DISCUSSION THREAD

**Discussion by beldmit** [openssl/openssl#25126] [General Discussion]
A follow-up to this discussion: https://www.openwall.com/lists/oss-security/2024/08/06/1

I have a different proposal. Can the provider API be expanded to move the TLS 1.0/1.1 protocol implementation either to a legacy provider or to some new provider? It seems to be a preferable way to deal with it.

**Comment by t8m** [comment]
That would be a huge undertaking.

**Reply by beldmit** [reply]
I agree that it's a huge undertaking. The question is whether it is worth it or not.

**Reply by Sashan** [reply]
If I understand things right the provider API allows us to plug additional cryptographic modules (ciphers, hashes, etc.), but not protocols. I would not mix those two things.

In my opinion we should not make it easier to keep those legacy protocols alive. In an ideal world, this should be disabled at compile time by default before it gets removed completely. The barrier to enabling TLS 1.0/1.1 should be high enough to prevent careless users from shooting their foot by enabling this legacy stuff with a simple tweak in the `.cnf` file.

**Reply by nhorman** [reply]
What represents a "high enough" barrier to entry though?

From the discussion here:
https://www.openwall.com/lists/oss-security/2024/08/06/1

The take aways I'm seeing are that most people are ok with removal as long as their is _some way_ to re-enable it, and those people have every intention of doing whatever that re-enablement is, because they have no other choice to keep their product alive.

The general feeling I get is that a build time default off option is probably best here, with a big build time warning that appears if enable-deprecated-tls is set during build config (or whatever toggle we want to use).  End users will do whatever they need to to keep their product running, but developers will at least have to pause and consider a warning (especially if they're building with -Werror)

**Reply by Sashan** [reply]
> What represents a "high enough" barrier to entry though?

I would say: `./Configure enable-legacy-tls` is sufficiently high barrier for the start.

**Reply by paulidale** [reply]
There was an intention for pluggable (i.e. provider based) protocols.  Or at least significant parts which would include TLS cipher suites etc.  We're not there yet of course.

**Reply by rsbeckerca** [reply]
I am starting the consultation process with the HPE NonStop community on this. What might be helpful is splitting the conversation, if possible, into ciphers suites vs. TLS 1.0/1.1 - As an example, we have an ongoing problem with 1.3 clients not wanting to talk to 1.2 servers with different cipher suites, and I don't know how to address that one (partly because the 1.2 build is not under my control), but it has created some confusion that I would like to be able to clarify in simple terms - I'm not that good at speaking "Simply".
